### PR TITLE
Cassandra 16685 4.0

### PR DESCRIPTION
**Pull request by bereng**

### CODE REVIEW

**Comment by bereng** [comment]
- [J11](https://app.circleci.com/pipelines/github/bereng/cassandra/326/workflows/33f50b70-2ce7-400c-9b06-e1840a48450d/jobs/3115) repeated
- [J8](https://app.circleci.com/pipelines/github/bereng/cassandra/326/workflows/3f8816a9-0e74-4439-92b8-d8ef81bf2a97/jobs/3113) & [J8/11](https://app.circleci.com/pipelines/github/bereng/cassandra/326/workflows/3f8816a9-0e74-4439-92b8-d8ef81bf2a97/jobs/3114) repeated

File: test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java [hunk 1]
```diff
@@ -387,10 +407,13 @@ public void testRejectWhenPoolFullStrategy() throws InterruptedException
             {
                 // expected
             }
+
             // allow executing tests to complete
             blocked.signalAll();
-            completed.await(10, TimeUnit.SECONDS);
+            completed.await(11, TimeUnit.SECONDS);
```
Comment (michaelsembwever) [inline test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java]: nit: could we get a comment here please? or even better make `10` a well-named constant so we understand that's the wait happening inside the task.

File: test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java [hunk 2]
```diff
@@ -376,8 +384,20 @@ public void testRejectWhenPoolFullStrategy() throws InterruptedException
...
+
+            /*
+             * CASSANDRA-16685 When the underlying executor's queue is a SynchronousQueue, there can be races just after
+             * the ThreadPool's initialization while juggling and spinning up threads internally leading to false
+             * rejections. That queue needs a thread ready to pick up the task immediately or it will produce a reject
+             * exception upon 'offer()' method call on the executor's code. If the executor is still initializing or
+             * threads are not ready to take work you can get false rejections.
+             *
+             * A sleep has been added to give time to the thread pool to be ready to get work.
+             */
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java]: If this is a Java bug maybe we could explicitly mention it here

File: test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java [hunk 3]
```diff
@@ -387,10 +407,13 @@ public void testRejectWhenPoolFullStrategy() throws InterruptedException
...
                 // expected
             }
+
             // allow executing tests to complete
             blocked.signalAll();
-            completed.await(10, TimeUnit.SECONDS);
+            completed.await(11, TimeUnit.SECONDS);
+
             // Submission is unblocked
+            Thread.sleep(100);
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java]: Why do we need this second sleep?

File: test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java (hunk 1 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java]: +1

File: test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java (hunk 3 above)
Comment (bereng) [reply test/unit/org/apache/cassandra/service/ActiveRepairServiceTest.java]: Again the same error is produced if we don't wait. My guess being between task completion and TP readiness there is a small window the queue/TP interaction is broken.

**Comment by bereng** [comment]
Review comments addressed and I raised the timeout to 250 just to be on the safe side :shrug:

**Review by michaelsembwever — APPROVED** [review]

**Review by adelapena — APPROVED** [review]

**Comment by bereng** [comment]
Squashed & rebased
CI [j11](https://app.circleci.com/pipelines/github/bereng/cassandra/339/workflows/e2d8bff5-937a-4b1f-92ce-2f9e9f9a9554)
CI [j8](https://app.circleci.com/pipelines/github/bereng/cassandra/339/workflows/2ae5710a-4318-45f7-99c9-60feb5cd2284)

**Pull request closed without merging**